zebra-chain/src/orchard/keys.rs

```diff
@@ -564,18 +580,25 @@ impl From<FullViewingKey> for [u8; 96] {
     }
 }
 
-impl From<SpendingKey> for FullViewingKey {
+impl TryFrom<SpendingKey> for FullViewingKey {
+    type Error = &'static str;
     /// Derive a _full viewing key_ from a existing _spending key_.
     ///
     /// <https://zips.z.cash/protocol/nu5.pdf#addressesandkeys>
     /// <https://zips.z.cash/protocol/nu5.pdf#orchardfullviewingkeyencoding>
-    fn from(sk: SpendingKey) -> FullViewingKey {
-        let spend_authorizing_key = SpendAuthorizingKey::from(sk);
-
-        Self {
-            spend_validating_key: SpendValidatingKey::from(spend_authorizing_key),
-            nullifier_deriving_key: NullifierDerivingKey::from(sk),
-            ivk_commit_randomness: IvkCommitRandomness::from(sk),
+    fn try_from(sk: SpendingKey) -> Result<Self, Self::Error> {
+        if let Ok(spend_authorizing_key) = SpendAuthorizingKey::try_from(sk) {
+            if let Ok(spend_validation_key) = SpendValidatingKey::try_from(spend_authorizing_key) {
+                Ok(Self {
+                    spend_validating_key: spend_validation_key,
+                    nullifier_deriving_key: NullifierDerivingKey::from(sk),
+                    ivk_commit_randomness: IvkCommitRandomness::from(sk),
+                })
+            } else {
+                Err("Invalid spend validation key")
+            }
+        } else {
+            Err("Invalid spend authorizing key")
         }
     }
```

Comment:
Optional: Can this be simplified by using `?` to return the original errors?

Reply:
Nice one, i did this at https://github.com/ZcashFoundation/zebra/pull/4609/commits/8948f22aff823ff0b2738d1c80be32b545c7c796